sysfs/net_class.go

```diff
@@ -0,0 +1,127 @@
+	IfIndex          uint64 `fileName:"ifindex"`            // /sys/class/net/<iface>/ifindex
+	IfLink           uint64 `fileName:"iflink"`             // /sys/class/net/<iface>/iflink
+	LinkMode         uint64 `fileName:"link_mode"`          // /sys/class/net/<iface>/link_mode
+	Mtu              uint64 `fileName:"mtu"`                // /sys/class/net/<iface>/link_mode
+	NameAssignType   uint64 `fileName:"name_assign_type"`   // /sys/class/net/<iface>/name_assign_type
+	NetDevGroup      uint64 `fileName:"netdev_group"`       // /sys/class/net/<iface>/netdev_group
+	OperState        string `fileName:"operstate"`          // /sys/class/net/<iface>/operstate
+	PhysPortId       string `fileName:"phys_port_id"`       // /sys/class/net/<iface>/phys_port_id
+	PhysPortName     string `fileName:"phys_port_name"`     // /sys/class/net/<iface>/phys_port_name
+	PhysSwitchId     string `fileName:"phys_switch_id"`     // /sys/class/net/<iface>/phys_switch_id
+	Speed            uint64 `fileName:"speed"`              // /sys/class/net/<iface>/speed
+	TxQueueLen       uint64 `fileName:"tx_queue_len"`       // /sys/class/net/<iface>/tx_queue_len
+	Type             uint64 `fileName:"type"`               // /sys/class/net/<iface>/type
+}
+
+type NetClass map[string]interfaceNetClass
+
+// NewNetDev returns kernel/system statistics read from /proc/net/dev.
+func NewNetClass() (NetClass, error) {
+	fs, err := NewFS(DefaultMountPoint)
+	if err != nil {
+		return nil, err
+	}
+
+	return fs.NewNetClass()
+}
+
+func (fs FS) NewNetClass() (NetClass, error) {
+	return newNetClass(fs.Path("class/net"))
+}
+
+func newNetClass(path string) (NetClass, error) {
+	devices, err := ioutil.ReadDir(path)
+	if err != nil {
+		return NetClass{}, fmt.Errorf("cannot access %s dir %s", path, err)
+	}
+
+	netClass := NetClass{}
+	for _, deviceDir := range devices {
+		interfaceClass, err := netClass.parseInterfaceNetClass(path + "/" + deviceDir.Name())
+		if err != nil {
+			continue
+		}
+		interfaceClass.Name = deviceDir.Name()
+		netClass[deviceDir.Name()] = *interfaceClass
+	}
+	return netClass, nil
+}
+
+func (nc NetClass) parseInterfaceNetClass(devicePath string) (*interfaceNetClass, error) {
+	fields := []string{"AddrAssignType", "AddrLen", "Address", "Broadcast", "Carrier", "CarrierChanges", "CarrierUpCount", "CarrierDownCount", "DevId", "Dormant", "Duplex", "Flags", "IfAlias", "IfIndex", "IfLink", "LinkMode", "Mtu", "NameAssignType", "NetDevGroup", "OperState", "PhysPortId", "PhysPortName", "PhysSwitchId", "Speed", "TxQueueLen", "Type"}
+	interfaceClass := interfaceNetClass{}
+	interfaceElem := reflect.ValueOf(&interfaceClass).Elem()
+	interfaceType := reflect.TypeOf(interfaceClass)
+
+	for _, fieldName := range fields {
+		fieldType, found := interfaceType.FieldByName(fieldName)
+		if !found {
+			continue
+		}
+		fieldValue := interfaceElem.FieldByName(fieldName)
+
+		fileContents, err := ioutil.ReadFile(devicePath + "/" + fieldType.Tag.Get("fileName"))
```

Comment:
There should be a check whether fileName tag is set before trying to open such file.

Reply:
definitely. Added, thanks for that